vncviewer/DesktopWindow.cxx

```diff
@@ -850,50 +860,74 @@ int DesktopWindow::fltkHandle(int event, Fl_Window *win)
   return ret;
 }
 
-
 void DesktopWindow::fullscreen_on()
 {
-  if (not fullScreenAllMonitors)
-    fullscreen_screens(-1, -1, -1, -1);
-  else {
-    int top, bottom, left, right;
-    int top_y, bottom_y, left_x, right_x;
-
-    int sx, sy, sw, sh;
-
-    top = bottom = left = right = 0;
-
-    Fl::screen_xywh(sx, sy, sw, sh, 0);
-    top_y = sy;
-    bottom_y = sy + sh;
-    left_x = sx;
-    right_x = sx + sw;
-
-    for (int i = 1;i < Fl::screen_count();i++) {
-      Fl::screen_xywh(sx, sy, sw, sh, i);
-      if (sy < top_y) {
-        top = i;
-        top_y = sy;
-      }
-      if ((sy + sh) > bottom_y) {
-        bottom = i;
-        bottom_y = sy + sh;
-      }
-      if (sx < left_x) {
-        left = i;
-        left_x = sx;
-      }
-      if ((sx + sw) > right_x) {
-        right = i;
-        right_x = sx + sw;
+  int top, bottom, left, right;
+  top = bottom = left = right = -1;
+  bool all_monitors = !strcasecmp(fullScreenMode, "all");
+  bool selected_monitors = !strcasecmp(fullScreenMode, "selected");
+
+  if (selected_monitors || all_monitors) {
+      std::set<int> monitors;
+
+      if (selected_monitors && !all_monitors) {
+        std::set<int> selected = fullScreenSelectedMonitors.getParam();
+        monitors.insert(selected.begin(), selected.end());
+      } else {
+        for (int i = 0; i < Fl::screen_count(); i++) {
+          monitors.insert(i);
+        }
       }
-    }
 
-    fullscreen_screens(top, bottom, left, right);
+      // If there are monitors selected, calculate the dimensions
+      // of the frame buffer, expressed in the monitor indices that
+      // limits it. 
+      if (monitors.size() > 0) {
```

Comment:
Do we log selected but missing monitors at some point? That would be useful if users are getting problems.